barretenberg/cpp/src/barretenberg/vm/avm/trace/execution.cpp

```diff
@@ -258,585 +294,637 @@ bool Execution::verify(AvmFlavor::VerificationKey vk, HonkProof const& proof)
-        if (setup_requests.contract_address != 0) {
-            public_call_requests.push_back(setup_requests);
-        }
-    }
-    for (const auto& app_requests : public_inputs.public_app_logic_call_requests) {
-        if (app_requests.contract_address != 0) {
-            public_call_requests.push_back(app_requests);
-        }
+    const auto setup_call_requests = non_empty_call_requests(public_inputs.public_setup_call_requests);
+    const auto app_logic_call_requests = non_empty_call_requests(public_inputs.public_app_logic_call_requests);
+    std::vector<PublicCallRequest> teardown_call_requests;
+    if (!public_inputs.public_teardown_call_request.is_empty()) {
+        // teardown is always one call request
+        teardown_call_requests.push_back(public_inputs.public_teardown_call_request);
     }
-    // We should not need to guard teardown, but while we are testing with handcrafted txs we do
-    if (public_inputs.public_teardown_call_request.contract_address != 0) {
-        public_call_requests.push_back(public_inputs.public_teardown_call_request);
-    }
-
-    // We should use the public input address, but for now we just take the first element in the list
-    // const std::vector<uint8_t>& bytecode = execution_hints.all_contract_bytecode.at(0).bytecode;
 
     // Loop over all the public call requests
     uint8_t call_ctx = 0;
-    for (const auto& public_call_request : public_call_requests) {
-        trace_builder.set_public_call_request(public_call_request);
-        trace_builder.set_call_ptr(call_ctx++);
-
-        // Find the bytecode based on contract address of the public call request
-        const std::vector<uint8_t>& bytecode =
-            std::ranges::find_if(execution_hints.all_contract_bytecode, [public_call_request](const auto& contract) {
-                return contract.contract_instance.address == public_call_request.contract_address;
-            })->bytecode;
-        info("Found bytecode for contract address: ", public_call_request.contract_address);
-
-        // Set this also on nested call
-
-        // Copied version of pc maintained in trace builder. The value of pc is evolving based
-        // on opcode logic and therefore is not maintained here. However, the next opcode in the execution
-        // is determined by this value which require read access to the code below.
-        uint32_t pc = 0;
-        uint32_t counter = 0;
-        AvmError error = AvmError::NO_ERROR;
-        while (is_ok(error) && (pc = trace_builder.get_pc()) < bytecode.size()) {
-            auto [inst, parse_error] = Deserialization::parse(bytecode, pc);
-            error = parse_error;
-
-            if (!is_ok(error)) {
+    const auto phases = { TxExecutionPhase::SETUP, TxExecutionPhase::APP_LOGIC, TxExecutionPhase::TEARDOWN };
+    for (auto phase : phases) {
+        const auto public_call_requests = phase == TxExecutionPhase::SETUP       ? setup_call_requests
```

Comment:
I think you create a routine for the code which is after "if ... else if" (Let us name it: process_req()) and you could turn the code as:

// code related to SETUP //
process_req(setup_call_requests)
// code related to APP_LOGIC //
process_req(app_logic_call_requests)
process_req(tear_down_call_requests)